src/swap/handle_swap_sign_transaction.c

```
@@ -0,0 +1,219 @@
+        PRINTF("Amount validated in swap = %s\n", validated_amount_str);
+        return false;
+    }
+    return true;
+}
+
+/**
+ * Validates that the transaction parameters saved in the app memory
+ * are the same as the ones in the Exchange App current G_context.
+ *
+ * @return true if success,false otherwise.
+ *
+ */
+bool swap_check_validity() {
+    PRINTF("Inside Aptos swap_check_validity\n");
+    if (!G_swap_validated.initialized) {
+        PRINTF("Swap Validated data not initialized.\n");
+        return false;
+    }
+    // Validate it's and actual coin transfer type
+    if (G_context.tx_info.transaction.tx_variant != TX_RAW) {
+        PRINTF("TX variant different from TX_RAW is not compatible with Swap.\n");
+        return false;
+    }
+
+    PRINTF("Payload variant: %d\n", G_context.tx_info.transaction.payload_variant);
+    if (G_context.tx_info.transaction.payload_variant != PAYLOAD_ENTRY_FUNCTION) {
+        PRINTF("Payload variant incompatible with Swap.\n");
+        return false;
+    }
+
+    // Differentiate between the different types of transactions
+    uint64_t amount = 0;
+    uint8_t* receiver;
+    switch (G_context.tx_info.transaction.payload.entry_function.known_type) {
+        case FUNC_APTOS_ACCOUNT_TRANSFER:
+            amount = G_context.tx_info.transaction.payload.entry_function.args.transfer.amount;
+            receiver = G_context.tx_info.transaction.payload.entry_function.args.transfer.receiver;
+            break;
+        case FUNC_COIN_TRANSFER:
+        case FUNC_APTOS_ACCOUNT_TRANSFER_COINS:
+            amount = G_context.tx_info.transaction.payload.entry_function.args.coin_transfer.amount;
+            receiver =
+                G_context.tx_info.transaction.payload.entry_function.args.coin_transfer.receiver;
+            break;
+        default:
+            PRINTF("Unknown function type\n");
+            return false;
+    }
+
+    if (!validate_swap_amount(amount)) {
```

If we are swapping tokens, do we check that we are trading the token validated in exchange ?

---

In this case, shouldn't we return an error if `G_context.tx_info.transaction.payload.entry_function.known_type` is `FUNC_COIN_TRANSFER`/`FUNC_APTOS_ACCOUNT_TRANSFER_COINS`? The flow is unsupported and the clear signing will be incorrect.

---

Indeed, I read the documentation again and looks good to me. I though these functions could handle tokens.